File: leetcode/questions/221/python/221.py
from typing import List

class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        rows, cols = len(matrix), len(matrix[0])
        aux = [[0 for _ in range(cols + 1)] for _ in range(rows + 1)]
        ans = 0

        for row in range(1, rows + 1):
            for col in range(1, cols + 1):
                if matrix[row - 1][col - 1] == "1":
                    aux[row][col] = min(aux[row - 1][col], aux[row][col - 1], aux[row - 1][col - 1]) + 1

                ans = max(ans, aux[row][col])

        return ans * ans

File: leetcode/questions/221/java/221.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[][] aux = new int[rows + 1][cols + 1];
        int len = 0;

        for (int row = 1; row <= rows; ++row) {
            for (int col = 1; col <= cols; ++col) {
                if (matrix[row - 1][col - 1] == '1') {
                    aux[row][col] = Math.min(aux[row - 1][col], Math.min(aux[row][col - 1], aux[row - 1][col - 1])) + 1;
                }
                len = Math.max(len, aux[row][col]);
            }
        }

        return len * len;
    }
}